### ISSUE TITLE
i915.enable_gvt broke suspend/hibernate

### ISSUE THREAD

**Issue by franz82** [01org/gvt-linux#156]
Hi all,
I'm trying to fix suspend/hibernate on my host with vfio enabled. Everything worked fine (I was able to suspend and hibernate the host), I don't know which update (maybe kernel since v5.6.3) broke it.
If  I remove i915.enable_gvt=1 from boot options, suspend/hibernate works fine.
If I boot with i915.enable_gvt=1, if I suspend/hibernate the host is freezed, screen is on, ssh session to the notebook hang, fan is working, nothing helpful on screen ([I opened a thread in Arch forum](https://bbs.archlinux.org/viewtopic.php?id=254777) with some info).

**Host**
I'm running a X1 Carbon 7th Gen with a Intel(R) Core(TM) i5-8265U CPU with Arch Linux.

```
$ uname -a
Linux x1c7 5.7.4-arch1-1 #1 SMP PREEMPT Thu, 18 Jun 2020 16:01:07 +0000 x86_64 GNU/Linux

$ virsh version
Compiled against library: libvirt 6.4.0
Using library: libvirt 6.4.0
Using API: QEMU 6.4.0
Running hypervisor: QEMU 5.0.0
```

Modules and boot options:

```
$ cat /etc/modules-load.d/kvm.conf 
kvm
kvm_intel
virtio
virtio_pci
virtio_net
virtio_blk
virtio_ring

$ cat /etc/default/grub        
...
GRUB_CMDLINE_LINUX="cryptdevice=/dev/nvme0n1p3:luks:allow-discards i915.enable_gvt=1 resume=/dev/vg0/swap"
```

**Guest**
Guest is a simple Win10 VM, it works fine - dump of the VM config below:

```
$ sudo virsh dumpxml win10
setlocale: No such file or directory
<domain type='kvm' xmlns:qemu='http://libvirt.org/schemas/domain/qemu/1.0'>
  <name>win10</name>
  <uuid>911333f7-0a27-4466-8f4d-2127e0a9fef0</uuid>
  <metadata>
    <libosinfo:libosinfo xmlns:libosinfo="http://libosinfo.org/xmlns/libvirt/domain/1.0">
      <libosinfo:os id="http://microsoft.com/win/10"/>
    </libosinfo:libosinfo>
  </metadata>
  <memory unit='KiB'>8388608</memory>
  <currentMemory unit='KiB'>8388608</currentMemory>
  <vcpu placement='static'>2</vcpu>
  <cputune>
    <vcpupin vcpu='0' cpuset='1'/>
    <vcpupin vcpu='1' cpuset='2'/>
  </cputune>
  <os>
    <type arch='x86_64' machine='pc-q35-4.1'>hvm</type>
  </os>
  <features>
    <acpi/>
    <apic/>
    <hyperv>
      <relaxed state='on'/>
      <vapic state='on'/>
      <spinlocks state='on' retries='8191'/>
    </hyperv>
    <vmport state='off'/>
  </features>
  <cpu mode='host-passthrough' check='none'>
    <topology sockets='1' dies='1' cores='1' threads='2'/>
  </cpu>
  <clock offset='localtime'>
    <timer name='hpet' present='yes'/>
    <timer name='hypervclock' present='yes'/>
  </clock>
  <on_poweroff>destroy</on_poweroff>
  <on_reboot>restart</on_reboot>
  <on_crash>destroy</on_crash>
  <pm>
    <suspend-to-mem enabled='no'/>
    <suspend-to-disk enabled='no'/>
  </pm>
  <devices>
    <emulator>/usr/bin/qemu-system-x86_64</emulator>
    <disk type='file' device='cdrom'>
      <driver name='qemu' type='raw'/>
      <source file='/home/franz/Downloads/Win10_1903_V1_Italian_x64.iso'/>
      <target dev='sdb' bus='sata'/>
      <readonly/>
      <boot order='1'/>
      <address type='drive' controller='0' bus='0' target='0' unit='1'/>
    </disk>
    <disk type='file' device='cdrom'>
      <driver name='qemu' type='raw'/>
      <source file='/home/franz/Downloads/virtio-win-0.1.141.iso'/>
      <target dev='sdc' bus='sata'/>
      <readonly/>
      <address type='drive' controller='0' bus='0' target='0' unit='2'/>
    </disk>
    <disk type='file' device='disk'>
      <driver name='qemu' type='qcow2'/>
      <source file='/var/lib/libvirt/images/win10.qcow2'/>
      <target dev='vda' bus='virtio'/>
      <boot order='2'/>
      <address type='pci' domain='0x0000' bus='0x04' slot='0x00' function='0x0'/>
    </disk>
    <controller type='usb' index='0' model='qemu-xhci' ports='15'>
      <address type='pci' domain='0x0000' bus='0x02' slot='0x00' function='0x0'/>
    </controller>
    <controller type='sata' index='0'>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x1f' function='0x2'/>
    </controller>
    <controller type='pci' index='0' model='pcie-root'/>
    <controller type='pci' index='1' model='pcie-root-port'>
      <model name='pcie-root-port'/>
      <target chassis='1' port='0x10'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x02' function='0x0' multifunction='on'/>
    </controller>
    <controller type='pci' index='2' model='pcie-root-port'>
      <model name='pcie-root-port'/>
      <target chassis='2' port='0x11'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x02' function='0x1'/>
    </controller>
    <controller type='pci' index='3' model='pcie-root-port'>
      <model name='pcie-root-port'/>
      <target chassis='3' port='0x12'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x02' function='0x2'/>
    </controller>
    <controller type='pci' index='4' model='pcie-root-port'>
      <model name='pcie-root-port'/>
      <target chassis='4' port='0x13'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x02' function='0x3'/>
    </controller>
    <controller type='pci' index='5' model='pcie-root-port'>
      <model name='pcie-root-port'/>
      <target chassis='5' port='0x14'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x02' function='0x4'/>
    </controller>
    <controller type='pci' index='6' model='pcie-root-port'>
      <model name='pcie-root-port'/>
      <target chassis='6' port='0x15'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x02' function='0x5'/>
    </controller>
    <controller type='pci' index='7' model='pcie-root-port'>
      <model name='pcie-root-port'/>
      <target chassis='7' port='0x16'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x02' function='0x6'/>
    </controller>
    <controller type='pci' index='8' model='pcie-root-port'>
      <model name='pcie-root-port'/>
      <target chassis='8' port='0x17'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x02' function='0x7'/>
    </controller>
    <controller type='pci' index='9' model='pcie-root-port'>
      <model name='pcie-root-port'/>
      <target chassis='9' port='0x18'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x03' function='0x0'/>
    </controller>
    <controller type='virtio-serial' index='0'>
      <address type='pci' domain='0x0000' bus='0x03' slot='0x00' function='0x0'/>
    </controller>
    <interface type='network'>
      <mac address='52:54:00:8c:14:25'/>
      <source network='default'/>
      <model type='e1000e'/>
      <address type='pci' domain='0x0000' bus='0x01' slot='0x00' function='0x0'/>
    </interface>
    <serial type='pty'>
      <target type='isa-serial' port='0'>
        <model name='isa-serial'/>
      </target>
    </serial>
    <console type='pty'>
      <target type='serial' port='0'/>
    </console>
    <channel type='spicevmc'>
      <target type='virtio' name='com.redhat.spice.0'/>
      <address type='virtio-serial' controller='0' bus='0' port='1'/>
    </channel>
    <input type='mouse' bus='ps2'/>
    <input type='keyboard' bus='ps2'/>
    <graphics type='spice'>
      <listen type='none'/>
      <image compression='off'/>
      <streaming mode='off'/>
      <gl enable='yes' rendernode='/dev/dri/by-path/pci-0000:00:02.0-render'/>
    </graphics>
    <sound model='ich9'>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x1b' function='0x0'/>
    </sound>
    <video>
      <model type='none'/>
    </video>
    <hostdev mode='subsystem' type='mdev' managed='no' model='vfio-pci' display='on'>
      <source>
        <address uuid='db2a0a3b-f3b3-4851-b459-f17bef11483d'/>
      </source>
      <address type='pci' domain='0x0000' bus='0x09' slot='0x00' function='0x0'/>
    </hostdev>
    <memballoon model='virtio'>
      <address type='pci' domain='0x0000' bus='0x05' slot='0x00' function='0x0'/>
    </memballoon>
  </devices>
  <qemu:commandline>
    <qemu:arg value='-set'/>
    <qemu:arg value='device.hostdev0.x-igd-opregion=on'/>
    <qemu:arg value='-set'/>
    <qemu:arg value='device.hostdev0.display=on'/>
    <qemu:arg value='-set'/>
    <qemu:arg value='device.hostdev0.ramfb=on'/>
    <qemu:arg value='-set'/>
    <qemu:arg value='device.hostdev0.driver=vfio-pci-nohotplug'/>
    <qemu:env name='MESA_LOADER_DRIVER_OVERRIDE' value='i965'/>
  </qemu:commandline>
</domain>
```

Thanks for any help,
-f

**Comment by ymarkus** [comment]
Having the same issue here. It is also fixed if I remove all my vGPUs. 

**Host**
X1 Carbon 6th Gen with Intel(R) Core(TM) i5-8350U CPU
```
[user@hostname:~]$ uname -a
Linux hostname 5.7.2 #1-NixOS SMP Wed Jun 10 18:21:45 UTC 2020 x86_64 GNU/Linux
```

**Comment by emilkhatib** [comment]
Hi! Having the same problem here too. I just started using GVT, so I haven't had experience with previous kernels.

**Host**
Thinkpad L380 with i7-8550U

**OS**
OpenSuse Tumbleweed. 
`
user@hostname:~> uname -a
Linux hostname 5.7.5-1-default #1 SMP Tue Jun 23 06:00:46 UTC 2020 (a1775d0) x86_64 x86_64 x86_64 GNU/Linux
`

**Comment by bartraw** [comment]
same issue on t560 with Intel(R) Core(TM) i7-6600U CPU @ 2.60GHz

**Comment by franz82** [comment]
As workaround I created another grub entry with i915.enable_gvt option enabled, removing it in the default one.
When I need to use VMs I choose the right grub entry at boot.
-f

**Comment by mazzz1y** [comment]
Same for me, i5-10210u on MSI Prestige 14

**Comment by chriscjs** [comment]
I am having the same issue, and I have tried the latest gvt-staging.  Host cannot enter suspend/hibernate if any vGPUs exist.

**Comment by offlinehacker** [comment]
I am having same issue. Host crashes on suspend/hibernate if any vGPU exists.

This makes gvt-g unusable on linux workstation machines, where you need S3/S4. gvt-g is currently only feature why I still prefer intel over AMD on workstations, so this would be really nice to have fixed. It's also a good selling point for intel, as all alternative vgpu solutions are too expensive.

**Host**
Dell inspiron 7391 with Core i7-10510U

```
Linux hostname 5.7.6 #1-NixOS SMP Wed Jun 24 15:49:26 UTC 2020 x86_64 GNU/Linux
```

**Comment by chriscjs** [comment]
I opened an issue about this a few months ago here
https://gitlab.freedesktop.org/drm/intel/-/issues/1633

I recently traced this bug down to the following commit:
drm/i915: Remove i915->kernel_context
https://cgit.freedesktop.org/drm-intel/commit/?h=drm-intel-next-2019-12-23&id=e6ba76480299a0d77c51d846f7467b1673aad25b

which was merged into linux 5.6 early on.

The git bisect log is in the last [post.](https://gitlab.freedesktop.org/drm/intel/-/issues/1633#note_601942)

**Comment by oakszyjrnrdy** [comment]
Same for me, Intel i7-9850H, Intel UHD Graphics 630 and NVIDIA GeForce GTX 1650 Mobile / Max-Q using prime offloading.

**Comment by oakszyjrnrdy** [comment]
A workaround for me is to remove the GVT-g virtual GPU before suspend and recreate it after waking from suspend. I wrote a [systemd file](https://github.com/oakszyjrnrdy/gvtg_vgpu) to help me automatically do this.

**Comment by TerrenceXu** [comment]
Thanks all, we will track this issue and provide the feedback to you all.

**Comment by chriscjs** [comment]
This issue is fixed in latest gvt-staging.  Need to put VM in s3/s4 or powered off before suspending host.
Commit https://github.com/intel/gvt-linux/commit/401ccfa87856656b874c737522ea92721394a348

**Comment by nikp123** [comment]
@chriscjs Sorry to bother, but how can (or does) QEMU/XEN handle S3/S4 states?
I thought you were supposed to disable that in order to get GVT-g working at all.

**Comment by chriscjs** [comment]
There were commits earlier this year that allow s3/s4 in VM with GVT-g.
See https://github.com/intel/gvt-linux/commit/9e7c0efadb86ddb58965561bbca638d44792d78f and https://github.com/intel/gvt-linux/commit/ba25d977571e1551b7032d6104e49efd6f88f8ad
Not sure which kernel you are using, but I am using 5.9 and those commits are already upstreamed.

**Comment by nikp123** [comment]
> There were commits earlier this year that allow s3/s4 in VM with GVT-g.
> See [9e7c0ef](https://github.com/intel/gvt-linux/commit/9e7c0efadb86ddb58965561bbca638d44792d78f) and [ba25d97](https://github.com/intel/gvt-linux/commit/ba25d977571e1551b7032d6104e49efd6f88f8ad)
> Not sure which kernel you are using, but I am using 5.9 and those commits are already upstreamed.

Ok, thanks, I wasnt up on the news.

**Comment by chriscjs** [comment]
With these patches, included in kernel 5.11, the host hangs when the VM with vgpu is resumed from sleep after the host is resumed from sleep.  In kernel 5.4 lts, Windows VM's recover gracefully, so its back to 5.4....
